docs/BLSSignatureChecker.md

```diff
@@ -1 +1,196 @@
+[core-dmgr-deregister]: https://github.com/Layr-Labs/eigenlayer-contracts/blob/m2-mainnet/docs/core/DelegationManager.md#deregisteroperatorfromavs
+
+[eigenda-service-manager]: https://github.com/Layr-Labs/eigenda/blob/m2-mainnet-contracts/contracts/src/core/EigenDAServiceManager.sol
+
+## BLSSignatureChecker
+
+| File | Type | Proxy |
+| -------- | -------- | -------- |
+| [`BLSSignatureChecker.sol`](../src/BLSSignatureChecker.sol) | Singleton | Transparent proxy |
+| [`OperatorStateRetriever.sol`](../src/OperatorStateRetriever.sol) | Singleton | None |
+
+`BLSSignatureChecker` and `OperatorStateRetriever` perform (respectively) the onchain and offchain portions of BLS signature validation for the aggregate of a quorum's registered Operators.
+
+The `OperatorStateRetriever` has various view methods intended to be called by offchain infrastructure in order to prepare a call to `BLSSignatureChecker.checkSignatures`. These methods traverse the state histories kept by the various registry contracts (see [./RegistryCoordinator.md](./RegistryCoordinator.md)) to query states at specific block numbers.
+
+These historical states are then used within `BLSSignatureChecker` to validate a BLS signature formed from an aggregated subset of the Operators registered for one or more quorums at some specific block number.
+
+#### High-level Concepts
+
+This document organizes methods according to the following themes (click each to be taken to the relevant section):
+* [Onchain](#onchain)
+* [Offchain](#offchain)
+* [System Configuration](#system-configuration)
+
+---
+
+### Onchain
+
+#### `BLSSignatureChecker.checkSignatures`
+
+```solidity
+function checkSignatures(
+    bytes32 msgHash, 
+    bytes calldata quorumNumbers,
+    uint32 referenceBlockNumber, 
+    NonSignerStakesAndSignature memory params
+)
+    public 
+    view
+    returns (QuorumStakeTotals memory, bytes32)
+
+struct NonSignerStakesAndSignature {
+    uint32[] nonSignerQuorumBitmapIndices;
+    BN254.G1Point[] nonSignerPubkeys;
+    BN254.G1Point[] quorumApks;
+    BN254.G2Point apkG2;
+    BN254.G1Point sigma;
+    uint32[] quorumApkIndices;
+    uint32[] totalStakeIndices;
+    uint32[][] nonSignerStakeIndices;
+}
+
+struct QuorumStakeTotals {
+    uint96[] signedStakeForQuorum;
+    uint96[] totalStakeForQuorum;
+}
+```
+
+The goal of this method is to allow an AVS to validate a BLS signature formed from the aggregate pubkey ("apk") of Operators registered in one or more quorums at some `referenceBlockNumber`.
+
+Some notes on method parameters:
+* `referenceBlockNumber` is the reason each registry contract keeps historical states: so that lookups can be performed on each registry's info at a particular block. This is important because Operators may sign some data on behalf of an AVS, then deregister from one or more of the AVS's quorums. Historical states allow signature validation to be performed against a "fixed point" in AVS/quorum history.
```

Comment:
this also helps deal with race conditions. if all you keep track of is the current APK, then when a new operator registered, it would cause all "in-flight" confirmations to revert.

Reply:
ah, on re-reading, maybe that's what you're already saying 😄